consensus/istanbul/backend/handler.go

```diff
@@ -290,3 +298,191 @@ func (sb *Backend) UnregisterPeer(peer consensus.Peer, isProxiedPeer bool) {
 		}
 	}
 }
+
+// Handshake allows the initiating peer to identify itself as a validator
+func (sb *Backend) Handshake(peer consensus.Peer, peerIsInbound bool) (bool, error) {
+	// Only written to if there was a non-nil error when sending or receiving
+	errCh := make(chan error, 1)
+	isValidatorCh := make(chan bool, 1)
+
+	sendHandshake := func() {
+		validatorProofMessage, containsProof, err := sb.generateValidatorProofMessage(peer)
+		if err != nil {
+			errCh <- err
+			return
+		}
+		msgBytes, err := validatorProofMessage.Payload()
+		if err != nil {
+			errCh <- err
+			return
+		}
+		err = peer.Send(istanbulValidatorProofMsg, msgBytes)
+		if err != nil {
+			errCh <- err
+		}
+		isValidatorCh <- containsProof
+	}
+	readHandshake := func() {
+		isValidator, err := sb.readValidatorProofMessage(peer)
+		if err != nil {
+			errCh <- err
+			return
+		}
+		isValidatorCh <- isValidator
+	}
+
+	// Only the initating peer sends the message
+	if peerIsInbound {
+		go readHandshake()
+	} else {
+		go sendHandshake()
+	}
+
+	timeout := time.NewTimer(handshakeTimeout)
+	defer timeout.Stop()
+	for {
+		select {
+		case err := <-errCh:
+			return false, err
+		case <-timeout.C:
+			return false, p2p.DiscReadTimeout
+		case isValidator := <-isValidatorCh:
+			return isValidator, nil
+		}
+	}
+}
+
+// generateValidatorProofMessage will create a message that contains a proof
+// to the peer during the handshake that this node is a validator, which is just
+// a versioned enode message.
+// If this node is not a validator or this node does not believe the peer
+// is a validator, no proof is generated to hide the address of this node
+// and an empty message is created.
+// If this node is a proxy, it will use the most recent versioned enode message this
+// node has received from the proxied validator if one exists.
+// Returns the message to send, if it contains a proof this node is a validator,
+// and if an error occurred.
+func (sb *Backend) generateValidatorProofMessage(peer consensus.Peer) (*istanbul.Message, bool, error) {
+	shouldSend, err := sb.shouldSendValidatorProof(peer)
+	if err != nil {
+		return nil, false, err
+	}
+	if shouldSend {
+		msg, err := sb.retrieveSelfVersionedEnodeMsg(getCurrentAnnounceVersion())
+		if err != nil {
+			return nil, false, err
+		}
+		if msg != nil {
+			return msg, true, nil
+		}
+	}
+	// Even if we decide not to identify ourselves,
+	// send an empty message to complete the handshake
+	return &istanbul.Message{}, false, nil
+}
+
+// shouldSendValidatorProof determines if this node should send a
+// validator proof revealing its address to a peer
+func (sb *Backend) shouldSendValidatorProof(peer consensus.Peer) (bool, error) {
```

Comment:
If you can figure out a way to do this:  https://github.com/celo-org/celo-blockchain/pull/873/files#r385459074, then this check could become unnecessary.

Reply:
Also, keep in mind that the proof should be sent only when a validator connection (which is created in the istanbul package and you should be able to figure out that it's a validator connection) is initiated.

I also echo @nategraf 's concern in the comment below, so I think it would be a big value-add if that can be avoided all together.